File: Book/ch14_SortPrac/실패율.js
const n1 = 5;
const stages1 = [2, 1, 2, 6, 2, 4, 3, 3];
const n2 = 4;
const stages2 = [4, 4, 4, 4, 4];

const solution = (n, stages) => {
  const result = [];
  for (let current = 1; current <= n; current++) {
    const stay = stages.filter((v) => v === current).length;
    const all = stages.filter((v) => v >= current).length;
    const failure = stay / all;
    if (all === 0) failure = 0;
    result.push({ stage: current, failure: failure });
  }
  return result.sort((a, b) => b.failure - a.failure).map((v) => v.stage);
};

console.log(solution(n2, stages2));

File: Programmers/Level1/실패율.py
def solution(N, stages):
    answer = []
    length = len(stages)

    for stage in range(1, N + 1):
        stay = stages.count(stage)
        if length == 0:
            failure = 0
        else:
            failure = stay / length

        answer.append((stage, failure))
        length -= stay

    answer = sorted(answer, key=lambda x: x[1], reverse=True)
    return [i[0] for i in answer]

print(solution(5, [2, 1, 2, 6, 2, 4, 3, 3]))